src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -3793,9 +3793,70 @@ private static void NativeCompletionMemberName(string wordToComplete, List<Compl
             }
 
             CompleteMemberByInferredType(context, prevType, result, wordToComplete + "*", filter: IsPropertyMember, isStatic: false);
+            if (result.Count == 0)
+            {
+                // if we get here, we have not been able to infer the type to complete.
+                // try looking at the value of the closest commandexpression upstream
+                CommandExpressionAst commandExpression = null;
+                for (int pipelineElementIndex = i - 1; pipelineElementIndex >= 0 && commandExpression == null; --pipelineElementIndex)
+                {
+                    var command = pipelineAst.PipelineElements[pipelineElementIndex] as CommandAst;
+                    // skip where,sort etc since they don't change they type of the pipeline object
+                    if (command != null && !IsCompletionTypeInvariantCommand(command))
+                    {
+                        break;
+                    }
+                    commandExpression = pipelineAst.PipelineElements[pipelineElementIndex] as CommandExpressionAst;
+                }
+                if (commandExpression != null)
+                {
+                    var expr = commandExpression.Expression;
+                    // if expr is $p.Where{...}, use $p
+                    if (IsCompletionTypeInvariantInvokeMemberExpression(expr))
+                    {
+                        expr = ((InvokeMemberExpressionAst)expr).Expression;
+                    }
+                    object value;
+                    SafeExprEvaluator.TrySafeEval(expr, context.ExecutionContext, out value);
+                    if (value != null)
+                    {
+                        var list = value as IList;
+                        if (list != null && list.Count > 0)
+                        {
+                            value = list[0];
+                        }
+                        value = PSObject.Base(value);
+                        CompleteMemberByActualType(context, new PSTypeName(value.GetType()), result, wordToComplete + "*", filter: IsPropertyMember, isStatic: false);
+                    }
+                }
+            }
             result.Add(CompletionResult.Null);
         }
 
+        static bool IsCompletionTypeInvariantCommand(CommandAst commandAst)
+        {
+            var commandName = commandAst.CommandElements[0].Extent.Text;
+            if (string.Compare(commandName, "where", StringComparison.OrdinalIgnoreCase) == 0) return true;
```

Comment:
For faster comparisons, you should use the string that most likely appears in the script, which is usually whatever tab completion returns, so `Where-Object` instead of `where-object`.